apps/state/lib/state/connecting_stops.ex

```diff
@@ -0,0 +1,192 @@
+  end
+
+  @doc "Force a synchronous state update, for use in tests."
+  def update!(overrides \\ @overrides) do
+    :ok = GenServer.call(__MODULE__, {:update!, overrides})
+  end
+
+  @impl Events.Server
+  def handle_event(_, _, _, state) do
+    {:noreply, maybe_update(state), :hibernate}
+  end
+
+  @impl GenServer
+  def init(_) do
+    @table = :ets.new(@table, [:named_table, read_concurrency: true])
+    Events.publish({:new_state, __MODULE__}, 0)
+    for server <- @sources, do: Events.subscribe({:new_state, server})
+    {:ok, maybe_update(nil), :hibernate}
+  end
+
+  @impl GenServer
+  def handle_call({:update!, overrides}, _from, state) do
+    {:reply, :ok, update_state(state, overrides), :hibernate}
+  end
+
+  @spec maybe_update(state) :: state
+  defp maybe_update(state) do
+    # Normally sources should only be empty if they are still initializing their own state, so in
+    # that case we shouldn't update ours yet
+    if Enum.all?(@sources, &(&1.size > 0)), do: update_state(state), else: state
+  end
+
+  @spec update_state(state, overrides) :: state
+  defp update_state(state, overrides \\ @overrides) do
+    items =
+      State.Stop.by_location_type([1])
+      |> Stream.filter(&Stop.located?/1)
+      |> Stream.map(fn stop -> [stop | connecting_stops(stop, overrides)] end)
+      |> Stream.flat_map(&rotations/1)
+      |> group_and_merge_by_head()
+      |> Stream.map(fn {stop, stops} -> {stop, reject_redundant_connections(stop, stops)} end)
+      |> Enum.map(fn {stop, stops} -> {stop.id, Enum.map(stops, & &1.id)} end)
+
+    :ets.delete_all_objects(@table)
+    :ets.insert(@table, items)
+    size = length(items)
+
+    _ =
+      Logger.info(fn ->
+        "Update #{__MODULE__} #{inspect(self())}: #{size} items"
+      end)
+
+    publish({:new_state, __MODULE__}, size)
+    state
+  end
+
+  # Find possible connecting stops for a given stop, taking overrides into account.
+  @spec connecting_stops(Stop.t(), overrides) :: [Stop.t()]
+  defp connecting_stops(stop, overrides) do
+    standalone_stops_around(stop)
+    |> MapSet.new()
+    |> MapSet.union(MapSet.new(override_stops(overrides, stop, :add)))
+    |> MapSet.difference(MapSet.new(override_stops(overrides, stop, :remove)))
+    |> MapSet.to_list()
+  end
+
+  # Find "standalone" stops close to a given stop's entrances, if it has entrances, otherwise
+  # close to the stop itself.
+  @spec standalone_stops_around(Stop.t()) :: [Stop.t()]
+  defp standalone_stops_around(%{id: stop_id, latitude: latitude, longitude: longitude}) do
+    case entrances(stop_id) do
+      [] ->
+        State.Stop.around(latitude, longitude, @radius) |> Enum.filter(&Stop.standalone?/1)
+
+      entrances ->
+        entrances |> Stream.flat_map(&standalone_stops_around/1) |> Enum.uniq()
+    end
+  end
+
+  @spec entrances(Stop.id()) :: [Stop.t()]
+  defp entrances(parent_station) do
+    State.Stop.Cache.match(%{location_type: 2, parent_station: parent_station}, :parent_station)
```

Comment:
Does this need to use the `Cache`, vs calling `State.Stop.match`? `Cache` seems like an implementation detail that this shouldn't depend on.

Reply:
`State.Stop.match` doesn't currently exist. Should it? Or should there be like a `State.Stop.by_location_type_and_parent_station`?

Reply:
Either way is good; I don't think you should talk directly to the Cache :)